### PR TITLE
[rocfft][cufft] DFT update host task to use native command

### PR DESCRIPTION
**Pull request by JackAKirk**
# Description

Similar to https://github.com/oneapi-src/oneMKL/pull/572 (see the discussions in that PR for technical details) except this covers fft backends for both amd and nvidia cases.

Update host task impl to use enqueue_native_command for DFT using the cuda/hip backends.

tests:

[test_main_dft_ct_amd.txt](https://github.com/user-attachments/files/17233320/test_main_dft_ct_amd.txt)
[test_main_dft_rt_amd.txt](https://github.com/user-attachments/files/17233321/test_main_dft_rt_amd.txt)
[test_main_dft_rt_nvidia.txt](https://github.com/user-attachments/files/17233322/test_main_dft_rt_nvidia.txt)
[test_main_dft_ct_nvidia.txt](https://github.com/user-attachments/files/17233323/test_main_dft_ct_nvidia.txt)

author: @hjabird

### CODE REVIEW

**Comment by JackAKirk** [comment]
@oneapi-src/onemkl-dft-write could you please review this?

Thanks

File: src/dft/backends/rocfft/execute_helper.hpp
Comment (Rbiessy) [inline src/dft/backends/rocfft/execute_helper.hpp]:
```suggestion
    auto result = rocfft_execute(plan, in_buffer, out_buffer, info);
```

**Review by Rbiessy — APPROVED** [review]
Thanks for the PR! Note we have used this in 2 GROMACS benchmarks to get between 2% to 7% improvements on MI210.

**Comment by JackAKirk** [comment]
@anantsrivastava30 Is this OK to merge?

Thanks

File: src/dft/backends/cufft/execute_helper.hpp
Comment (lhuot) [inline src/dft/backends/cufft/execute_helper.hpp:161]: Why is it needed to duplicate this in both cuFFT and rocFFT backends and in various domains (BLAS, LAPACK, FFT)? Can't we have one wrapper used across all domains and backends?

File: src/dft/backends/cufft/execute_helper.hpp
Comment (lhuot) [inline src/dft/backends/cufft/execute_helper.hpp]: Why is it needed to duplicate this in both cuFFT and rocFFT backends and in various domains (BLAS, LAPACK, FFT)? Can't we have one wrapper used across all domains and backends?

**Comment by JackAKirk** [comment]
Hi @lhuot 

I don't know, I am following the existing design. I think that it is a good question and it would be good to open this as an issue. However I don't think it is a good idea to block this PR for this, since this is a critical patch for Gromacs performance, which does not introduce this duplication that already exists in all other backends.

File: src/dft/backends/cufft/execute_helper.hpp
Comment (Rbiessy) [reply src/dft/backends/cufft/execute_helper.hpp]: My understanding is that the domains have always been separated on purpose. This makes review much easier as any change affecting common code would technically require an approval from every domain owners.
I agree this could be discussed in an issue. To my knowledge there is very little code that could be common across domains, other than the types and exceptions which are already common.

**Comment by lhuot** [comment]
> Hi @lhuot
> 
> I don't know, I am following the existing design. I think that it is a good question and it would be good to open this as an issue. However I don't think it is a good idea to block this PR for this, since this is a critical patch for Gromacs performance, which does not introduce this duplication that already exists in all other backends.

Then, let's fix the code duplication in the DFT domain please.

**Comment by JackAKirk** [comment]
> > Hi @lhuot
> > I don't know, I am following the existing design. I think that it is a good question and it would be good to open this as an issue. However I don't think it is a good idea to block this PR for this, since this is a critical patch for Gromacs performance, which does not introduce this duplication that already exists in all other backends.
> 
> Then, let's fix the code duplication in the DFT domain please.

OK sure, do you want me to add a new header e.g. "execute_helper_generic.hpp" in oneMKL/src/dft/ with the fft_host_task implementation that is portable across hip and cuda backends, and then only include this header in the rocfft and cufft backends?
I can't include such a generic header in include/oneapi/mkl/detail/ because the `enqueue_native_command` extension is not available in intel backends and trying to link this header will break intel builds.

**Comment by lhuot** [comment]
> > > Hi @lhuot
> > > I don't know, I am following the existing design. I think that it is a good question and it would be good to open this as an issue. However I don't think it is a good idea to block this PR for this, since this is a critical patch for Gromacs performance, which does not introduce this duplication that already exists in all other backends.
> > 
> > 
> > Then, let's fix the code duplication in the DFT domain please.
> 
> OK sure, do you want me to add a new header e.g. "execute_helper_generic.hpp" in oneMKL/src/dft/ with the fft_host_task implementation that is portable across hip and cuda backends, and then only include this header in the rocfft and cufft backends? I can't include such a generic header in include/oneapi/mkl/detail/ because the `enqueue_native_command` extension is not available in intel backends and trying to link this header will break intel builds.

Sounds like a reasonable approach to me. Thanks!

File: src/dft/execute_helper_generic.hpp
Comment (Rbiessy) [inline src/dft/execute_helper_generic.hpp]: `_ONEMKL_DFT_SRC_CUFFT_EXECUTE_GENERIC_HPP_` -> `_ONEMKL_DFT_SRC_EXECUTE_HELPER_GENERIC_HPP_`

The `endif` should also have a comment with the same name (looks like it was missed in a few headers).

Comment (Rbiessy) [reply src/dft/execute_helper_generic.hpp]: Sorry for the nitpicking, it should really end with `EXECUTE_HELPER_GENERIC_HPP_` not just `EXECUTE_GENERIC_HPP_`!

**Review by Rbiessy — APPROVED** [review]

**Comment by JackAKirk** [comment]
> > > > Hi @lhuot
> > > > I don't know, I am following the existing design. I think that it is a good question and it would be good to open this as an issue. However I don't think it is a good idea to block this PR for this, since this is a critical patch for Gromacs performance, which does not introduce this duplication that already exists in all other backends.
> > > 
> > > 
> > > Then, let's fix the code duplication in the DFT domain please.
> > 
> > 
> > OK sure, do you want me to add a new header e.g. "execute_helper_generic.hpp" in oneMKL/src/dft/ with the fft_host_task implementation that is portable across hip and cuda backends, and then only include this header in the rocfft and cufft backends? I can't include such a generic header in include/oneapi/mkl/detail/ because the `enqueue_native_command` extension is not available in intel backends and trying to link this header will break intel builds.
> 
> Sounds like a reasonable approach to me. Thanks!

I've updated this now as requested. Tests still pass as posted in PR summary on both hip and cuda backends.

**Review by lhuot — APPROVED** [review]
LGTM, thanks!

**Pull request merged**
